plugins/events-node/src/service.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  coreServices,
+  createServiceFactory,
+  createServiceRef,
+} from '@backstage/backend-plugin-api';
+import { EventsService, DefaultEventsService } from './api';
+
+/**
+ * The {@link EventsService} that allows to publish events, and subscribe to topics.
+ * Uses the `root` scope so that events can be shared across all plugins, modules, and more.
+ *
+ * @public
+ */
+export const eventsServiceRef = createServiceRef<EventsService>({
+  id: 'events.service',
+  scope: 'root',
```

Comment:
We were looking into merging this for the mainline release, but I'm sorry we think there's one thing that still should be settled.

We were wondering if we could

- make this one plugin scoped instead
- make it depend on the plugin metadata service and prefixing subscription IDs with `<pluginId>.`

Then we could possibly introduce a `rootEventsServiceRef` as well in the future but only if strictly needed (and reimplement the plugin scoped one in terms of the root scoped one). This also makes the naming pattern of this one more aligned.

Does that make sense? We usually add the root ones only if necessary since they kind of by definition are less "powerful". Do you have any particular reason to have a root scoped events service already at this point?

Happy to hear your feedback

Reply:
@freben The events service is supposed to create an event-based communication across plugin boundaries. Hence, I fail to see how it can be plugin-scoped.

If just the instance itself is plugin-scoped the actual underlying communication is still root-scoped/global, it could work. Even though the benefits might be limited (e.g., the prefixing of subscription IDs).

I guess the `version-bridge` was an idea to create this shared state/context, despite being plugin-scoped. However, this package is limited to `web-library` roles.

We could have both, root-scoped and plugin-scoped, and the root-scoped service is used within the plugin-scoped implementations. Not sure if this would help with what you intend to achieve.

Reply:
Yeah you create the shared state [with `createRootContext`](https://github.com/backstage/backstage/blob/eecadfad195eb5d821b5f04856bb2f88a89620bd/packages/backend-app-api/src/services/implementations/database/databaseServiceFactory.ts#L32) if you need one.

This choice isn't really based on what communication pattern it implements under the hood. It's based on the potential capabilities and whether it might want to contextually be aware of which plugin is using it. And you often do in the end; imagine wanting to emit plugin-tagged metrics based on their events activity for just one example among many.

Reply:
Thanks for the pointer, forgot about that one. I will have a look at how I can change the setup further.

Reply:
I switched to plugin-scope. Please let me know if this version works for you.